src/serialization/gpsw_ser.c

```diff
@@ -0,0 +1,258 @@
+    cfe_vec_octet_init(&octets, a->t.size);
+    for (size_t i =0; i<a->t.size; i++) {
+        char *h1 = (char *) cfe_malloc((4 * MODBYTES_256_56) * sizeof(char));
+        octet oct = {0, (4 * MODBYTES_256_56) * sizeof(char), h1};
+        ECP2_BN254_toOctet(&oct, &(a->t.vec[i]));
+        oct.max = oct.len;
+        octets.vec[i] = oct;
+    }
+    cfe_vec_octet_pack(&octets, msg->t, octets_val);
+
+    for (size_t i =0; i<a->t.size; i++) {
+        free(octets.vec[i].val);
+    }
+    free(octets.vec);
+}
+
+void cfe_gpsw_pub_key_ser(cfe_gpsw_pub_key *a, cfe_ser *buf) {
+    CfeGpswPubKeySer msg = CFE_GPSW_PUB_KEY_SER__INIT;
+    VecOctetSer t = VEC_OCTET_SER__INIT;
+    msg.t = &t;
+    OctetSer y = OCTET_SER__INIT;
+    msg.y = &y;
+    OctetSer *octets_val = cfe_malloc(sizeof(OctetSer) * a->t.size);
+
+    cfe_gpsw_pub_key_pack(a, &msg, octets_val);
+
+    buf->len = cfe_gpsw_pub_key_ser__get_packed_size(&msg);
+    buf->ser = cfe_malloc(buf->len);
+
+    cfe_gpsw_pub_key_ser__pack(&msg, buf->ser);
+
+    for (size_t i =0; i<a->t.size; i++) {
+        free(msg.t->vec[i]->val);
+    }
+    free(msg.t->vec);
+    free(msg.y->val);
+    free(octets_val);
+}
+
+void cfe_gpsw_pub_key_unpack(cfe_gpsw_pub_key *a, CfeGpswPubKeySer *msg) {
+    cfe_FP12_BN254_unpack(&(a->y), msg->y);
+
+    cfe_vec_octet octets;
+    cfe_vec_octet_unpack(&octets, msg->t);
+    cfe_vec_G2_init(&(a->t), octets.size);
+    for (size_t i =0; i< octets.size; i++) {
+        ECP2_BN254_fromOctet(&(a->t.vec[i]), &(octets.vec[i]));
+        free(octets.vec[i].val);
+    }
+    cfe_vec_octet_free(&octets);
+}
+
+cfe_error cfe_gpsw_pub_key_read(cfe_gpsw_pub_key *a, cfe_ser *buf) {
+    CfeGpswPubKeySer *msg;
+    msg = cfe_gpsw_pub_key_ser__unpack(NULL, buf->len, buf->ser);
+    if (msg == NULL) {
+        return 1;
+    }
+
+    cfe_gpsw_pub_key_unpack(a, msg);
+    // Free the unpacked message
+    cfe_gpsw_pub_key_ser__free_unpacked(msg, NULL);
+    return 0;
+}
+
+void cfe_gpsw_key_pack(cfe_gpsw_key *a, CfeGpswKeySer *msg, MpzSer *val, OctetSer *octets_val) {
+    cfe_msp_pack(&(a->msp), msg->msp, val);
+
+    cfe_vec_octet octets;
+    cfe_vec_octet_init(&octets, a->d.size);
+    for (size_t i =0; i<a->d.size; i++) {
+        char *h1 = (char *) cfe_malloc((MODBYTES_256_56 + 1) * sizeof(char));
+        octet oct = {0, (MODBYTES_256_56 + 1) * sizeof(char), h1};
+        ECP_BN254_toOctet(&oct, &(a->d.vec[i]), true);
+        oct.max = oct.len;
+        octets.vec[i] = oct;
+    }
+    cfe_vec_octet_pack(&octets, msg->d, octets_val);
+
+    for (size_t i =0; i<a->d.size; i++) {
+        free(octets.vec[i].val);
+    }
+    free(octets.vec);
+}
+
+void cfe_gpsw_key_ser(cfe_gpsw_key *a, cfe_ser *buf) {
+    CfeGpswKeySer msg = CFE_GPSW_KEY_SER__INIT;
+
+    MspSer msp = MSP_SER__INIT;
+    MatSer mat = MAT_SER__INIT;
+    msp.mat = &mat;
+
+    MpzSer *val = cfe_malloc(sizeof(MpzSer) * a->msp.mat.cols * a->msp.mat.rows);
+    cfe_msp_pack(&(a->msp), &msp, val);
```

Comment:
Valgrind is reporting lost blocks that are malloc'ed in this call:
```
==976020== 48 bytes in 1 blocks are definitely lost in loss record 2 of 4
==976020==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==976020==    by 0x4864F2C: cfe_malloc (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x48786B9: cfe_msp_pack (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x487936B: cfe_gpsw_key_ser (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x122A38: test_gpsw_key_ser (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x11F06B: munit_test_runner_exec (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x11FC3F: munit_test_runner_run_test_with_params (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12010B: munit_test_runner_run_test (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12028D: munit_test_runner_run_suite (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x1202C9: munit_test_runner_run_suite (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12093C: munit_suite_main_custom (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x114755: main (in /home/work/CiFEr/build/cifer_test)
==976020== 
==976020== 144 bytes in 1 blocks are definitely lost in loss record 3 of 4
==976020==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==976020==    by 0x4864F2C: cfe_malloc (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x4877CD4: cfe_mat_pack (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x48786A8: cfe_msp_pack (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x487936B: cfe_gpsw_key_ser (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x122A38: test_gpsw_key_ser (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x11F06B: munit_test_runner_exec (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x11FC3F: munit_test_runner_run_test_with_params (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12010B: munit_test_runner_run_test (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12028D: munit_test_runner_run_suite (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x1202C9: munit_test_runner_run_suite (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12093C: munit_suite_main_custom (in /home/work/CiFEr/build/cifer_test)
==976020== 
==976020== 288 bytes in 18 blocks are definitely lost in loss record 4 of 4
==976020==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==976020==    by 0x4864F2C: cfe_malloc (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x4877B14: cfe_mpz_pack (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x4877D47: cfe_mat_pack (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x48786A8: cfe_msp_pack (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x487936B: cfe_gpsw_key_ser (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x122A38: test_gpsw_key_ser (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x11F06B: munit_test_runner_exec (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x11FC3F: munit_test_runner_run_test_with_params (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12010B: munit_test_runner_run_test (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12028D: munit_test_runner_run_suite (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x1202C9: munit_test_runner_run_suite (in /home/work/CiFEr/build/cifer_test)
```
but if I'm not wrong I am freeing the memory [below](https://github.com/fentec-project/CiFEr/pull/28/files#diff-c6c0ce6bae55c5587d7fefc7399f7ac194c9e9748403c6cae610b85d98079f32R136).
I am missing something but I looked too much this code to see anything :sweat_smile: